server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceStats.java

```diff
@@ -81,7 +94,15 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.field("computed_shard_movements", computedShardMovements);
         builder.humanReadableField("computation_time_in_millis", "computation_time", new TimeValue(cumulativeComputationTime));
         builder.humanReadableField("reconciliation_time_in_millis", "reconciliation_time", new TimeValue(cumulativeReconciliationTime));
+        builder.field("unassigned_shards", unassignedShards);
+        builder.field("total_allocations", totalAllocations);
+        builder.field("undesired_allocations", undesiredAllocations);
+        builder.field("undesired_allocations_fraction", undesiredAllocationsFraction());
         builder.endObject();
         return builder;
     }
+
+    public double undesiredAllocationsFraction() {
+        return (double) undesiredAllocations / totalAllocations;
```

Comment:
nit for versions before ADDITIONAL_DESIRED_BALANCE_RECONCILIATION_STATS, this will be -1/-1 = 1 = 100%. I do not think it is a problem, but if we do not want that, we could extend the logic to return 0% in case of negative values. Similar comment for `logUndesiredAllocationsMetrics()`.